module/zfs/zfs_vfsops.c

```diff
@@ -1888,6 +1904,7 @@ zfs_preumount(struct super_block *sb)
 
 	/* zfsvfs is NULL when zfs_domount fails during mount */
 	if (zfsvfs) {
+		zfs_unlinked_drain_stop_wait(zfsvfs);
```

Comment:
I couldn't figure out when zfs_preumount() is called.  Is it possible that we later decide not to unmount it, in which case how do we restart the unlinked_drain task?

Reply:
it's called in `zpl_kill_sb()`. That function is called by the Linux kernel on unmount so afaik it's not possible to later decide not to unmount.

Reply:
Sorry for the confusion, when `zfs_preumount()` is called all inodes and dentrys have **not** yet been freed as correctly described in the comment above `zfs_preumount()`.  The dentry and inode caches will be torn down immediately after `zfs_preumount()` exits when `zpl_kill_sb()` calls `kill_anon_super()`.

```c
static void
zpl_kill_sb(struct super_block *sb)
{
        zfs_preumount(sb);
        kill_anon_super(sb);
}
```